src/builder.rs

```diff
@@ -372,98 +442,124 @@ impl Builder {
-                .into_iter()
-                .zip(self.recipients.into_iter())
-                .map(|(spend, recipient)| ActionInfo::new(spend, recipient, &mut rng))
-                .collect()
-        };
+        self,
+        rng: impl RngCore,
+    ) -> Result<Option<Bundle<InProgress<Unproven, Unauthorized>, V>>, BuildError> {
+        bundle(rng, self.spends, self.outputs, self.bundle_type)
+    }
+}
 
-        // Move some things out of self that we will need.
-        let flags = self.flags;
-        let anchor = self.anchor;
+/// Builds a bundle containing the given spent notes and outputs.
+///
+/// The returned bundle will have no proof or signatures; these can be applied with
+/// [`Bundle::create_proof`] and [`Bundle::apply_signatures`] respectively.
+pub fn bundle<V: TryFrom<i64>>(
+    mut rng: impl RngCore,
+    mut spends: Vec<SpendInfo>,
+    mut outputs: Vec<OutputInfo>,
+    bundle_type: BundleType,
+) -> Result<Option<Bundle<InProgress<Unproven, Unauthorized>, V>>, BuildError> {
+    let (flags, anchor) = bundle_type.bundle_config();
+
+    let num_requested_spends = spends.len();
+    if !flags.spends_enabled() && num_requested_spends > 0 {
+        return Err(BuildError::SpendsDisabled);
+    }
 
-        // Determine the value balance for this bundle, ensuring it is valid.
-        let value_balance = pre_actions
-            .iter()
-            .fold(Some(ValueSum::zero()), |acc, action| {
-                acc? + action.value_sum()
-            })
-            .ok_or(OverflowError)?;
+    for spend in &spends {
+        if !spend.has_matching_anchor(anchor) {
+            return Err(BuildError::AnchorMismatch);
+        }
+    }
 
-        let result_value_balance: V = i64::try_from(value_balance)
-            .map_err(BuildError::ValueSum)
-            .and_then(|i| V::try_from(i).map_err(|_| BuildError::ValueSum(value::OverflowError)))?;
+    let num_requested_outputs = outputs.len();
+    if !flags.outputs_enabled() && num_requested_outputs > 0 {
+        return Err(BuildError::OutputsDisabled);
+    }
 
-        // Compute the transaction binding signing key.
-        let bsk = pre_actions
-            .iter()
-            .map(|a| &a.rcv)
-            .sum::<ValueCommitTrapdoor>()
-            .into_bsk();
-
-        // Create the actions.
-        let (actions, circuits): (Vec<_>, Vec<_>) =
-            pre_actions.into_iter().map(|a| a.build(&mut rng)).unzip();
-
-        // Verify that bsk and bvk are consistent.
-        let bvk = (actions.iter().map(|a| a.cv_net()).sum::<ValueCommitment>()
-            - ValueCommitment::derive(value_balance, ValueCommitTrapdoor::zero()))
-        .into_bvk();
-        assert_eq!(redpallas::VerificationKey::from(&bsk), bvk);
-
-        Ok(Bundle::from_parts(
-            NonEmpty::from_vec(actions).unwrap(),
+    let num_actions = bundle_type
+        .num_actions(num_requested_spends, num_requested_outputs)
+        .map_err(|_| BuildError::BundleTypeNotSatisfiable)?;
```

Comment:
This is now a bit unintuitive, because it can potentially include both real spends and outputs, and dummies beyond what the bundle type's padding requires. This opens up the API to user-defined extra padding that can potentially make their transactions distinguishable from network transactions.

That being said, the way that `BundleType::num_actions` works means that this doesn't prevent the `BundleType`'s padding rule from being applied, so this is a hazard that should be documented, not a bug per se.

Reply:
The users can already add padding by just sending zeros to arbitrary addresses, so it doesn't change anything.